### PR TITLE
[memprof] Improve deserialization performance in V3

### PR DESCRIPTION
**Pull request by kazutakahirata**
We call llvm::sort in a couple of places in the V3 encoding:

- We sort Frames by FrameIds for stability of the output.

- We sort call stacks in the dictionary order to maximize the length
  of the common prefix between adjacent call stacks.

It turns out that we can improve the deserialization performance by
modifying the comparison functions -- without changing the format at
all.  Both places take advantage of the histogram of Frames -- how
many times each Frame occurs in the call stacks.

- Frames: We serialize popular Frames in the descending order of
  popularity for improved cache locality.  For two equally popular
  Frames, we break a tie by serializing one that tends to appear
  earlier in call stacks.  Here, "earlier" means a smaller index
  within llvm::SmallVector<FrameId>.

- Call Stacks: We sort the call stacks to reduce the number of times
  we follow pointers to parents during deserialization.  Specifically,
  instead of comparing two call stacks in the strcmp style -- integer
  comparisons of FrameIds, we compare two FrameIds F1 and F2 with
  Histogram[F1] < Histogram[F2] at respective indexes.  Since we
  encode from the end of the sorted list of call stacks, we tend to
  encode popular call stacks first.

Since the two places use the same histogram, we compute it once and
share it in the two places.

Sorting the call stacks reduces the number of "jumps" by 74% when we
deserialize all MemProfRecords.  The cycle and instruction counts go
down by 10% and 1.5%, respectively.

If we sort the Frames in addition to the call stacks, then the cycle
and instruction counts go down by 14% and 1.6%, respectively, relative
to the same baseline (that is, without this patch).

### CODE REVIEW

File: llvm/lib/ProfileData/InstrProfWriter.cpp [hunk 1]
```diff
@@ -494,17 +494,43 @@ static uint64_t writeMemProfFrames(
...
+    llvm::MapVector<memprof::FrameId, memprof::Frame> &MemProfFrameData,
+    llvm::DenseMap<memprof::FrameId, memprof::FrameStat> &FrameHistogram) {
   // Mappings from FrameIds to array indexes.
   llvm::DenseMap<memprof::FrameId, memprof::LinearFrameId> MemProfFrameIndexes;
 
-  // Sort the FrameIDs for stability.
+  // Compute the order in which we serialize Frames.  The order does not matter
+  // in terms of correctness, but we still compute it for two reasons:
+  //
+  // - Stability of the output: If we sort by FrameIds at least, then the output
```
Comment (teresajohnson) [inline llvm/lib/ProfileData/InstrProfWriter.cpp:505]: This is largely not done afaict? I guess as a fallback tie breaker, but maybe better to mention that briefly at the end of the next paragraph?

File: llvm/lib/ProfileData/MemProf.cpp [hunk 2]
```diff
@@ -499,12 +500,56 @@ void CallStackRadixTreeBuilder::build(
 
   // Sort the list of call stacks in the dictionary order to maximize the length
```
Comment (teresajohnson) [inline llvm/lib/ProfileData/MemProf.cpp:501]: Should this comment be removed?

**Review by teresajohnson — APPROVED** [review]
lgtm with a few minor suggestions

File: llvm/lib/ProfileData/InstrProfWriter.cpp (hunk 1 above)
Comment (kazutakahirata) [reply llvm/lib/ProfileData/InstrProfWriter.cpp]: Fixed in the latest iteration.  It's a lot better to mention the FrameId comparison as the final tie breaker all the away at the end.

File: llvm/lib/ProfileData/MemProf.cpp (hunk 2 above)
Comment (kazutakahirata) [reply llvm/lib/ProfileData/MemProf.cpp]: Fixed in the latest iteration.  I've removed the old comment and rephrased the immediately following paragraph.

**Pull request merged**